marimo/_plugins/ui/_impl/chat/chat.py

```diff
@@ -0,0 +1,175 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Any, Callable, Dict, Final, List, Optional
+
+from marimo._output.formatting import as_html
+from marimo._output.rich_help import mddoc
+from marimo._plugins.ui._core.ui_element import UIElement
+from marimo._plugins.ui._impl.chat.convert import model_from_callable
+from marimo._plugins.ui._impl.chat.types import (
+    ChatMessage,
+    ChatModel,
+    ChatModelConfig,
+    from_chat_message_dict,
+)
+from marimo._runtime.functions import EmptyArgs, Function
+
+
+@dataclass
+class SendMessageRequest:
+    messages: List[ChatMessage]
+    config: ChatModelConfig
+
+
+@dataclass
+class GetChatHistoryResponse:
+    messages: List[ChatMessage]
+
+
+DEFAULT_SYSTEM_MESSAGE = (
+    "You are a helpful assistant specializing in data science."
+)
+
+
+@mddoc
+class chat(UIElement[Dict[str, Any], List[ChatMessage]]):
+    """
+    A chatbot UI element for interactive conversations.
+
+    **Example - Using a custom model.**
+
+    You can define a custom chat model that takes a
+    prompt, previous messages, and optional attachments as input
+    and returns a response.
+
+    The response can be an object, a marimo UI element, or a string.
+
+    ```python
+    def my_rag_model(prompt, messages, attachments):
+        docs = find_docs(prompt)
+        prompt = template(prompt, docs, messages)
+        response = query(prompt)
+        if is_dataset(response):
+            return dataset_to_chart(response)
+        return response
+
+
+    chat = mo.ai.chatbot(
+        my_rag_model,
+        system_message="You are a helpful assistant.",
+    )
+    ```
+
+    **Example - Using a built-in model.**
+
+    You can use a built-in model from the `mo.ai` module.
+
+    ```python
+    chat = mo.ai.chatbot(
+        mo.ai.openai("gpt-4o"),
+        system_message="You are a helpful assistant.",
+    )
+    ```
+
+    **Attributes.**
+
+    - `value`: the current chat history
+
+    **Initialization Args.**
+
+    - `model`: the chatbot model
+    - `system_message`: the initial system message for the chatbot
+    - `prompts`: optional list of prompts to start the conversation
+    - `on_message`: optional callback function to handle new messages
+    - `max_tokens`: maximum number of tokens in the response
+    - `temperature`: sampling temperature for response generation
+    """
+
+    _name: Final[str] = "marimo-chatbot"
+
+    def __init__(
+        self,
+        model: ChatModel,
+        *,
+        system_message: str = DEFAULT_SYSTEM_MESSAGE,
+        prompts: Optional[List[str]] = None,
+        on_message: Optional[Callable[[List[ChatMessage]], None]] = None,
+        show_configuration_controls: bool = False,
+        max_tokens: int = 100,
+        temperature: float = 0.5,
+        top_p: float = 1,
+        top_k: int = 40,
+        frequency_penalty: float = 0,
+        presence_penalty: float = 0,
```

Comment:
Couldn't the configuration parameters used by models change over time, as new models are developed? Would it be possible to encapsulate the configuration in the `ChatModel` instance?

Reply:
I thought about this but then requires more duplication. No strong preference e